app/services/container_dashboard_service.rb

```diff
@@ -1,13 +1,15 @@
 class ContainerDashboardService
-  def initialize(provider_id)
+  def initialize(provider_id, controller)
     @provider_id = provider_id
     @ems = ManageIQ::Providers::ContainerManager.find(@provider_id) unless @provider_id.nil?
+    @controller = controller
   end
```

Comment:
To handle linking, I passed the controller to the service and used it to generate links to other entities.